[Feature] Support insert overwrite

ABingHuang opened the pull request:
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4798

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
Support insert overwrite for OlapTable. use as following:
insert overwrite tableA select xxx from tableB;
insert overwrite tableA partitions(p1, p2) select xxx from tableB;
The design idea is create mirror temporary partitions and load new data into the temporary partitions. when all data is loaded, swap temporary partitions and original partitions.
The core points are:
1. modify OlapTable meta of UNPARTITIONED table to support to add temporary partitions 
2. the atomicity of load. when the insert failed, should remove the temporary partitions created
3. the editlog and serialization to data persistence
4. the concurrent control between insert into and insert overwrite. now just mutual exclusion is used.

ABingHuang committed:
Support insert overwrite

ABingHuang committed:
fix compile

chaoyli commented:
Please change the title and obey the rule.

imay commented:
> Please change the title and obey the rule.

Can you give the rule's Link? I'm not sure where can I find it.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,409 @@
+        // should replan here because prepareInsert has changed the targetPartitionNames of insertStmt
+        ExecPlan newPlan = new StatementPlanner().plan(insertStmt, context);
+        stmtExecutor.handleDMLStmt(newPlan, insertStmt);
+        insertElapse = System.currentTimeMillis() - insertStartTimestamp;
```
insertElapse is unused?

ABingHuang replied:
it is used in log when job succeeds or fails

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,409 @@
+                AddPartitionsInfo infos = new AddPartitionsInfo(partitionInfoList);
+                GlobalStateMgr.getCurrentState().getEditLog().logAddPartitions(infos);
+            } finally {
+                createPartitionElapse = System.currentTimeMillis() - createPartitionStartTimestamp;
```
createPartitionElapse unused.

ABingHuang replied:
it is used in to to print time info.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,278 @@
+    private ReentrantReadWriteLock lock;
+
+    public InsertOverwriteJobManager() {
+        this.overwriteJobMap = Maps.newHashMap();
```
could we use concurrent hash map to avoid the explicit lock?

ABingHuang replied:
there are some related data structures should be modified atomically. So here I use lock.

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -231,7 +231,7 @@ dropFrontendClause
 
 insertStatement
-    : explainDesc? INSERT INTO qualifiedName partitionNames?
+    : explainDesc? INSERT (INTO | OVERWRITE) qualifiedName partitionNames?
```
If it is compatible with hive syntax, this should be insert (into | overwrite). However, there is another problem here. I think overwrite should be a reserve keyword, otherwise it cannot be distinguished from the following qualifiedName.

ABingHuang replied:
as [HangyuanLiu](https://github.com/HangyuanLiu) say, hive/spark use insert (INTO | OVERWRITE)，https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-dml-insert-overwrite-table.html

ABingHuang replied:
@HangyuanLiu ok, I will fix it

imay replied:
OK I think it is OK to compatible with Hive and spark.

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:
```diff
@@ -149,6 +149,7 @@ INTERMEDIATE: 'INTERMEDIATE';
 INTERSECT: 'INTERSECT';
 INTERVAL: 'INTERVAL';
 INTO: 'INTO';
+OVERWRITE: 'OVERWRITE';
```
lexicographically

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    public boolean hasRunningOverwriteJob(long txnId, long tableId, List<Long> partitions) {
```
Will we support insert into many partitions of one table at the same time?

ABingHuang replied:
yes, we do support this function.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java:
```diff
@@ -1382,6 +1388,7 @@ private void transferToMaster(FrontendNodeType oldType) {
             startMasterOnlyDaemonThreads();
             // start other daemon threads that should running on all FE
             startNonMasterDaemonThreads();
+            insertOverwriteJobManager.cancelRunningJobs();
```
why need to cancel running jobs when FE `transferToMaster`?

ABingHuang replied:
because I think the unfinished insert overwrite job should fail when master failover.

kangkaisen replied:
Why after master failover, the insert overwrite job should fail.  I think which load job, after master failover, we should continue run the job.

ABingHuang replied:
I test insert into, when master failover, it will fail. so I keep the same with it. And more, if we continue the job, we should make insert overwrite reenterable, which will need more state and is more complicated.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java:
```diff
@@ -13,7 +13,7 @@
 import com.starrocks.catalog.OlapTable;
 import com.starrocks.catalog.Partition;
-import com.starrocks.catalog.PartitionType;
+// import com.starrocks.catalog.PartitionType;
```
remove it

ABingHuang replied:
ok

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:
```diff
@@ -476,6 +476,13 @@ public void commitTransaction(long transactionId, List<TabletCommitInfo> tabletC
+            boolean hasRunningJobs =
+                    GlobalStateMgr.getCurrentState().getInsertOverwriteJobManager().hasRunningOverwriteJob(
+                            transactionId, tableId, tableToPartition.get(tableId).stream().collect(Collectors.toList()));
+            if (hasRunningJobs) {
```
should we check whether hasRunningJobs in here?

ABingHuang replied:
I find only here I can get the partitions to commit. So I check it here

kangkaisen replied:
I think this logic is a little strange. Why until one transaction is going to commit, we just check there running insert jobs.  I think we could check it early.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:
```diff
@@ -420,7 +423,38 @@ public void execute() throws Exception {
             } else if (parsedStmt instanceof DmlStmt) {
                 try {
-                    handleDMLStmt(execPlan, (DmlStmt) parsedStmt);
+                    if (parsedStmt instanceof InsertStmt && ((InsertStmt) parsedStmt).isOverwrite()) {
```
Abstract these following  codes to a method.

ABingHuang replied:
ok

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,278 @@
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class InsertOverwriteJobManager implements Writable, GsonPostProcessable {
```
if we have 1000 insert into overwrite jobs in FE, how many FE memories will use at most for the job metadata?

ABingHuang replied:
In my test, one InsertOverwriteJob will take up 56 bytes. and 1000 jobs will take up almost 56kbyes.

ABingHuang replied:
@kangkaisen PTAL

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,409 @@
+    private List<String> newPartitionNames;
+
+    @SerializedName(value = "dbId")
+    private long dbId;
```
targetDbId?

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,409 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class InsertOverwriteJob {
```
Can you split into two parts? A InsertOverwriteJobRunner and InsertOverwriteJobState.
The State will be persisted and the Runner would process a Job?

imay replied:
Then you decouple the State and the Runner. When users create an Job, you create an State to represent it. and then when you are going to execute it, you will create an Runner.

ABingHuang replied:
I split then original InsertOverwriteJob into two parts: InsertOverwriteJob and InsertOverwriteJobRunner. PTAL

ABingHuang replied:
@imay PTAL

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:
```diff
@@ -216,4 +216,6 @@ public class OperationType {
     // drop external catalog
     public static final short OP_DROP_CATALOG = 10061;
 
+    public static final short OP_CREATE_INSERT_OVERWRITE = 10070;
```
Do we have comments to explain the difference and connection between these two states ？ in the process of reviewing the code, I can't simply understand their meaning from the name

ABingHuang replied:
OP_CREATE_INSERT_OVERWRITE is for creation of insert overwrite job
OP_INSERT_OVERWRITE_STATE_CHANGE is for state change of insert overwrite job.

ABingHuang replied:
PTAL

ABingHuang committed:
fix comment

ABingHuang committed:
fix compile

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 166 / 569 (29.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|13|00.00%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|1|202|00.50%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|56|05.36%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|9|36|25.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|13|32|40.62%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|10|15|66.67%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|102|137|74.45%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|6|6|100.00%|

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,340 @@
+                doLoad();
+                break;
+            case OVERWRITE_FAILED:
+            case OVERWRITE_CANCELLED:
```
maybe use OVERWRITE_FAILED instead of OVERWRITE_CANCELLED will be better?

ABingHuang replied:
ok

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,340 @@
+        insertStmt.setOverwrite(false);
+        // should replan here because prepareInsert has changed the targetPartitionNames of insertStmt
+        ExecPlan newPlan = new StatementPlanner().plan(insertStmt, context);
+        stmtExecutor.handleDMLStmt(newPlan, insertStmt);
```
should we write audit.log after execute SQL?

ABingHuang replied:
sql has been added to audit.log

ABingHuang commented:
run starrocks_fe_unittest

ABingHuang commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 166 / 564 (29.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|13|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|38|00.00%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|1|202|00.50%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|9|36|25.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|13|32|40.62%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|10|15|66.67%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|102|137|74.45%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|6|6|100.00%|

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,340 @@
+            while (!isPreviousLoadFinished() && !context.isKilled()) {
+                Thread.sleep(500);
+            }
+            waitInsertIntoElapse = waitInsertStartTimestamp - waitInsertStartTimestamp;
```
both waitInsertStartTimestamp

ABingHuang replied:
fixed

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobState.java:
```diff
@@ -0,0 +1,11 @@
+
+public enum InsertOverwriteJobState {
+    OVERWRITE_PENDING,
+    OVERWRITE_PREPARED,
```
Why not have running state? what's difference between pending and prepared?

ABingHuang replied:
OVERWRITE_PREPARE is the same to OVERWRITE_RUNNING. I will change OVERWRITE_PREPARE to OVERWRITE_RUNNING.
OVERWRITE_PENDING is to get new partition names which are used to gc.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:
```diff
@@ -1022,6 +1025,32 @@ private boolean isStatisticsOrAnalyzer(StatementBase statement, ConnectContext c
     }
 
+    public void handleInsertOverwrite(InsertStmt insertStmt) {
+        Database database = MetaUtils.getStarRocks(context, insertStmt.getTableName());
```
`getStarRocks` should rename to getDatabase

ABingHuang committed:
fix comment

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 177 / 571 (31.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|13|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|38|00.00%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|1|196|00.51%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|11|40|27.50%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|13|32|40.62%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|10|15|66.67%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|102|137|74.45%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,289 @@
+            }
+            List<Long> runningPartitions = partitionsWithOverwrite.get(tableId);
+            if (runningPartitions == null || runningPartitions.isEmpty()) {
+                return true;
```
here should return true because when runningPartitions are null or empty means the table has running overwrite job

ABingHuang committed:
add ut and fix comment

ABingHuang committed:
fix compile

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 278 / 576 (48.26%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|0|40|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|13|15.38%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|49|196|25.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|4|12|33.33%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|10|15|66.67%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|12|17|70.59%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|102|137|74.45%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|32|90.62%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|

ABingHuang committed:
add ut

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 324 / 576 (56.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|49|196|25.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|12|50.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|26|40|65.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|102|137|74.45%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|32|90.62%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

ABingHuang committed:
fix comment

ABingHuang committed:
fix ut

ABingHuang commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 341 / 569 (59.93%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|61|195|31.28%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|12|50.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|26|40|65.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|106|131|80.92%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

ABingHuang committed:
add lock manager

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:
```diff
@@ -297,6 +297,23 @@ public long beginTransaction(List<Long> tableIdList, String label, TUniqueId req
 
             checkRunningTxnExceedLimit(sourceType);
 
+            // insert overwrite的时候需要避免lock
```
English

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/load/LockManager.java:
```diff
@@ -0,0 +1,249 @@
+import java.util.concurrent.locks.ReentrantLock;
+
+// non-reentrant lock manager
+public class LockManager {
```
Seems this class isn't general lock manager.

imay replied:
How about PartitionLockManager

imay commented on fe/fe-core/src/main/java/com/starrocks/load/LockManager.java:
```diff
@@ -0,0 +1,249 @@
+
+    private ReentrantLock lock;
+    // tableId -> LockType -> partitionIds
+    private Map<Long, Map<LockType, Set<PartitionEntity>>> lockTables;
```
why not `Map<partitionId, ReadWriteLock>`
I think it is not a good idea to use LockType as a Map key.

ABingHuang replied:
I will modify it

ABingHuang committed:
modify lock manager

ABingHuang committed:
Optimize LockManager

ABingHuang committed:
remove extra code

ABingHuang committed:
fix ut

ABingHuang commented:
run starrocks_fe_unittest

wangruin commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 501 / 819 (61.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|0|3|00.00%|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/load/lock/LockException.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|183|31.69%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|5|15|33.33%|
|:large_blue_circle:|com/starrocks/transaction/DatabaseTransactionMgr.java|23|61|37.70%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|9|19|47.37%|
|:large_blue_circle:|com/starrocks/load/lock/LockTarget.java|23|44|52.27%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|7|11|63.64%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/transaction/GlobalTransactionMgr.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|87|107|81.31%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/load/lock/LockTargetDesc.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/load/lock/LockManager.java|101|113|89.38%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/load/lock/Lock.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/load/lock/LockMode.java|3|3|100.00%|

ABingHuang committed:
optimize lock manager

ABingHuang committed:
fix compile

ABingHuang committed:
implement lock interface

ABingHuang committed:
modify level lock manager

ABingHuang committed:
remove level lock

ABingHuang committed:
remove code

ABingHuang committed:
fix schema change and rollup

ABingHuang committed:
fix code

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 321 / 543 (59.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|183|31.69%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|9|19|47.37%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|18|28|64.29%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|90|110|81.82%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 321 / 543 (59.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|183|31.69%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|10|40.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|9|19|47.37%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|18|28|64.29%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|90|110|81.82%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobInfo.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+
+    // tableId -> partitionId list
+    @SerializedName(value = "partitionsWithOverwrite")
+    private Map<Long, List<Long>> partitionsWithOverwrite;
```
Can you add more comments for this object?

ABingHuang replied:
it is modified to tableToOverwriteJobs

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+                    partitionsWithOverwrite.remove(job.getTargetTableId());
+                }
+            } else {
+                partitionsWithOverwrite.remove(job.getTargetTableId());
```
partitionIds is already null, why you execute remove ?

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+                partitionsWithOverwrite.remove(job.getTargetTableId());
+            }
+            overwriteJobMap.remove(jobid);
+            jobNum--;
```
seems equals with overwriteJobMap.size()

ABingHuang replied:
remove it

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    public void replayCreateInsertOverwrite(CreateInsertOverwriteJobInfo jobInfo) {
```
log and replay should in the same place.

ABingHuang replied:
done

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+
+    public void cancelRunningJobs() {
+        // resubmit running insert overwrite jobs
+        if (!GlobalStateMgr.isCheckpointThread()) {
```
if (GlobalStateMgr.isCheckpointThread()) return;

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+                            LOG.warn("InsertOverwriteJobManager runAfterCatalogReady interrupted exception.", e);
+                        }
+                    }
+                    if (runningJobs != null) {
```
needs lock to protect running jobs access.

ABingHuang replied:
runningJobs is only used for replay, so no need to add a lock

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+
+    private ExecutorService cancelJobExecutorService;
+
+    private List<InsertOverwriteJob> runningJobs;
```
give more comments

ABingHuang replied:
done

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,308 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class InsertOverwriteJobRunner {
```
provide more comments

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    private void createTempPartitions() {
```
Can you make it in a PartitionUtils class, then others can create tempartition easily. like
`PartitionUtils.createTempPartition`

ABingHuang replied:
done

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,308 @@
+                        .map(partitionName -> targetTable.getPartition(partitionName, true).getId())
+                        .collect(Collectors.toList());
+                PartitionNames partitionNames = new PartitionNames(true, job.getNewPartitionNames());
+                insertStmt.setOriginalTargetPartitionIds(insertStmt.getTargetPartitionIds());
+                insertStmt.setTargetPartitionNames(partitionNames);
+                insertStmt.setTargetPartitionIds(newPartitionIds);
```
why need db.lock here?

imay replied:
only one insert for all related partition?

ABingHuang replied:
I think targetTable.getPartition should be called in read lock

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java:
```diff
@@ -0,0 +1,308 @@
+        this.job = job;
+        this.context = context;
+        this.stmtExecutor = stmtExecutor;
+        this.insertStmt = job.getInsertStmt();
```
because you will change this insertStmt, you should deep copy it.

ABingHuang replied:
I just change the TargetPartitionNames and TargetPartitionIds, which are just some names and ids, generated by parser, and not something stored in meta. And I test that these fields do not impact the result of show process list. And more it seems not easy to deep copy InsertStmt now.
I think it is ok to modify some fields of InsertStmt.

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,104 @@
+    @SerializedName(value = "originalTargetPartitionIds")
+    private List<Long> originalTargetPartitionIds;
+
+    private InsertStmt insertStmt;
```
if the insertStmt is not serialized. how can the job rerun after FE restart.

ABingHuang replied:
Insert overwrite will be cancelled after FE restart, so do not need to serialize the insertStmt

imay commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJob.java:
```diff
@@ -0,0 +1,104 @@
+    private InsertOverwriteJobState jobState;
+
+    @SerializedName(value = "sourcePartitionNames")
+    private List<String> sourcePartitionNames;
```
if you already have partitionid, why you need partition names?

ABingHuang replied:
I modify it to use partition ids

imay commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -1670,6 +1670,31 @@ public void replaceTempPartitions(List<String> partitionNames, List<String> temp
 
+    // used for unpartitioned table in insert overwrite
+    // replace partition with temp partition
+    public void replacePartition(String sourcePartitionName, String tempPartitionName) {
```
It it strange to give partition names for unpartitioned

ABingHuang replied:
unpartitioned table's partition also has a name, it is default to the table name. Here I have to add a tmp partition for table, so I need to give the tmp partition a name. I think it is ok.

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/persist/CreateInsertOverwriteJobInfo.java:
```diff
@@ -0,0 +1,69 @@
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.List;
+public class CreateInsertOverwriteJobInfo implements Writable {
```
done

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobManager.java:
```diff
@@ -0,0 +1,239 @@
+        this.jobNum = 0;
+    }
+
+    public void submitJob(ConnectContext context, StmtExecutor stmtExecutor, InsertOverwriteJob job) throws Exception {
```
I will rename it to executeJob

ABingHuang committed:
fix comment

1. rename CreateInsertOverwriteJobInfo
2. use partition id to replace name
3. optimize log
4. extract partition utils

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 312 / 539 (57.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|0|34|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|157|36.94%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|43|53.49%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|14|24|58.33%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|91|111|81.98%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobLog.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

ABingHuang committed:
fix replay add partition

Astralidea commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 312 / 543 (57.46%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|0|34|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|157|36.94%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|49|46.94%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|14|24|58.33%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|91|111|81.98%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobLog.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

ABingHuang commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 312 / 543 (57.46%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|0|34|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|157|36.94%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|49|46.94%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|14|24|58.33%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|91|111|81.98%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobLog.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|

ABingHuang commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 312 / 543 (57.46%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/PartitionUtils.java|0|34|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|13|23.08%|
|:large_blue_circle:|com/starrocks/common/util/RangeUtils.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobRunner.java|58|157|36.94%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|23|49|46.94%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/analysis/InsertStmt.java|6|11|54.55%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|14|24|58.33%|
|:large_blue_circle:|com/starrocks/sql/analyzer/InsertAnalyzer.java|6|10|60.00%|
|:large_blue_circle:|com/starrocks/sql/common/MetaUtils.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobManager.java|91|111|81.98%|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|12|13|92.31%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJob.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/persist/CreateInsertOverwriteJobLog.java|14|15|93.33%|
|:large_blue_circle:|com/starrocks/persist/InsertOverwriteStateChangeInfo.java|16|17|94.12%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/load/InsertOverwriteJobState.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/DeleteAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzeStmtAnalyzer.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|6|6|100.00%|
